kernel/examples/table-reader/src/main/java/io/delta/kernel/examples/SingleThreadedTableReader.java

```diff
@@ -94,18 +100,37 @@ private int readData(StructType readSchema, Scan scan, int maxRowCount) throws I
 
         int readRecordCount = 0;
         try {
+            StructType physicalReadSchema =
+                ScanStateRow.getPhysicalDataReadSchema(tableClient, scanState);
             while (scanFileIter.hasNext()) {
-                try (CloseableIterator<FilteredColumnarBatch> data =
-                    Scan.readData(
-                        tableClient,
-                        scanState,
-                        scanFileIter.next().getRows(),
-                        Optional.empty())) {
-                    while (data.hasNext()) {
-                        FilteredColumnarBatch dataReadResult = data.next();
-                        readRecordCount += printData(dataReadResult, maxRowCount - readRecordCount);
-                        if (readRecordCount >= maxRowCount) {
-                            return readRecordCount;
+                FilteredColumnarBatch scanFilesBatch = scanFileIter.next();
+                try (CloseableIterator<Row> scanFileRows = scanFilesBatch.getRows()) {
+                    while (scanFileRows.hasNext()) {
+                        Row scanFileRow = scanFileRows.next();
+                        FileStatus fileStatus =
+                            InternalScanFileUtils.getAddFileStatus(scanFileRow);
```

Comment:
So there is a need for additional public utility methods to extract the params of `readParquetFiles` and `transformPhysicalData` from `scanStateRow` and `scanFileRow`. We should think carefully where to put these utility methods and make them future proof. For example, today physicalReadSchema depends only on `scanStateRow`, but tomorrow both `scanStateRow` and `scanFileRow` may dictate the schema. If we make the utility method take both params now, that will be way more future-proof.

Reply:
Based on the offline discussion with @tdas:

I think we need to start supporting utility methods to extract the info from the scan state `Row` and scan file `Row`. Given both of these (scan state and scan file) rows are generated by the `Scan`, is it better to add the following utility methods on `Scan` interface?

```
interface Scan {
  static StructType getPhysicalDataSchema(Row scanState, Row scanRow)

  static FileStatus getFileStatus(Row scanRow)
}
```

One concern with passing a `scanRow` to `getPhysicalDataSchema` is that it gets recomputed for every scan row. Without that we could just compute once for all scan files in a task. With it, the task need to make the call for every scan file row.

@ryan-johnson-databricks I think the above APIs are going to be useful for connectors to avoid reading through the schema of scan state and scan file rows and keeping the code updated as the schema of these contents change.

Reply:
Here `scanRow` represents a single add action? If so, that's not a very self-documenting name. 

@tdas it's a pretty significant architectural detail for the data schema to potentially change from file to file. For example, it means a lot of the choreography to turn the raw data from file into final logical data would have to occur on executors, which e.g. spark can't even do. It also means the engine expression that does the transformation has to be recreated on a per-file basis, because the schema has to be re-verified and re-adapted every time.

Can you elaborate on why we would want kernel to embrace that kind of complexity? What sort of Delta features would require it?

Reply:
For example, although field id handling requires such per-file schema awareness, we ended up pushing that into the parquet reader where it belongs.

Aside: We do need to figure out a way to cheaply pass per-file metadata to engine expressions, e.g. rowid base, so the engine doesn't have to create a new expression from scratch each time just to incorporate a new literal.

Reply:
I can open an issue and start the discussion once everybody is back from holidays. This PR doesn't need to block until we solve this issue?

Reply:
Doing schema stuff on a per-file vs. per-query basis is a pretty big decision. Can we move it out of this PR entirely, and just eliminate the contextualization?

Reply:
Agree, the schema stuff can be handled separately from this PR. Created https://github.com/delta-io/delta/issues/2395 to track this.